Show preview of Premium Content on frontend for admin

stacimc opened the pull request:
#### Changes proposed in this Pull Request

Currently, when the Premium Content block is not configured correctly -- either a plan upgrade is necessary, or a Stripe account needs to be connected -- it does not render on the frontend. This PR changes that behavior to instead display a preview of the block (the Logged Out View with the buttons disconnected) if the user viewing it is an admin.
This allows them to preview what the block will look like on the frontend of their site when they finish configuring it.

It is intended to be merged in conjunction with Automattic/jetpack#17793, which enables the same behavior for the Recurring Payments block.

The plan upgrade and Stripe connection nudges should also display in the frontend when relevant; the Stripe nudge will display without the `Connect` button since the url cannot be fully built in the SSR-rendered component. The user must return to the editor to connect to Stripe.

#### Testing instructions
Make sure you have Automattic/jetpack#17793 checked out as well. 

Should be tested using instructions from Automattic/jetpack#17793.

Commit:
Show preview of Premium Content on frontend for admin

When the Premium Content block is not configured correctly -- either
a plan upgrade is necessary, or a Stripe account needs to be connected
-- it does not render on the frontend. This commit changes this
behavior to instead display a preview of the block (the Logged Out
View with the buttons disconnected) if the user viewing it is an admin.
This allows them to preview what the block will look like on the
frontend of their site when they finish configuring it.

Commit:
Move isPremiumContentChild context to container

Commit:
Display upgrade and stripe nudges on frontend for admin

Commit:
Correctly register block context for Premium Content

This is used in the Recurring Payments block to determine whether
to render the button preview.

Commit:
Render the stripe connection nudge

Updates render callback to directly use render_component instead
of a custom render function for the stripe nudge.

Commit:
Render Stripe Nudge without connect button

Since the Stripe connect url cannot be fully built on the backend.
we pass null for the url to the nudge component so that no button
will be rendered. The user must return to the editor to connect.

apeatling commented on apps/editing-toolkit/editing-toolkit-plugin/premium-content/premium-content.php:
```diff
@@ -237,9 +300,66 @@ function premium_content_container_render( $attributes, $content ) {
+			'stripe-nudge',
+			array(
+				'blockName'        => 'premium-content',
+				'postId'           => $post_id,
```
This is not defined in the function so is not available. I think you'll want to use the `global $post;` at the top of the function, and use `$post->ID` instead.

stacimc replied:
Thanks -- casualty of a last minute refactor :)

apeatling reviewed (changes requested):
I'm seeing both the visitor and subscriber content rendered on the front end when running locally. I haven't yet connected Stripe, and I see the Stripe nudge displayed in the front end. I'll test this on my sandbox next.

<img width="681" alt="Screen Shot 2020-11-23 at 11 25 55 AM" src="https://user-images.githubusercontent.com/1464705/100005988-aa3a2a80-2d7e-11eb-8c8f-7cd963f2a731.png">

apeatling commented on apps/editing-toolkit/editing-toolkit-plugin/premium-content/premium-content.php:
```diff
@@ -237,9 +300,66 @@ function premium_content_container_render( $attributes, $content ) {
+
+	if ( function_exists( 'jetpack_require_lib' ) ) {
+		jetpack_require_lib( 'components' );
+		return Jetpack_Components::render_upgrade_nudge(
```
Need a `\` here before Jetpack_Components.

apeatling replied:
I think along with the `jetpack_require_lib` check, there should also be a `class_exists` check on the `Jetpack_Components` class. It's still possible even if the library loading code is present for some reason Jetpack_Components is not.

apeatling commented:
On my sandbox using a free plan, I saw the nudge, but I still see both the subscriber and visitor content when looking as an admin:

<img width="811" alt="Screen Shot 2020-11-23 at 11 42 47 AM" src="https://user-images.githubusercontent.com/1464705/100007548-19188300-2d81-11eb-940a-363fb036bffb.png">

Commit:
Do not render subscriber content in preview

stacimc commented:
> On my sandbox using a free plan, I saw the nudge, but I still see both the subscriber and visitor content when looking as an admin:
> 

Good catch, thanks -- I expect this would be showing up in all contexts. Fix is up to stop rendering the subscriber content when the 'preview' is being displayed.

apeatling reviewed (approved):
This is looking good now. :+1:

Commit:
Show preview if Stripe is connected but site needs upgrade

This is an edge case that can happen if a user signs up for a
premium (or higher) plan, connects a Stripe account, and then
removes the plan. Since the site requires the upgrade in order
to use the Premium Content block, we should still render the
preview in this scenario.

apeatling commented:
Going to close this one out since the changes are now merged into the move PR and have been tested. Thanks for your work on this @stacimc!
